File: python/arrays/longest_unique_subarray.py
# Example: Find the length of the longest subarray with all unique elements
# Duplicates make this problem harder because you must track and skip over repeated elements efficiently

def longest_unique_subarray(arr):
    max_len = 0
    start = 0
    seen = {}
    for end, num in enumerate(arr):
        if num in seen and seen[num] >= start:
            start = seen[num] + 1
        seen[num] = end
        max_len = max(max_len, end - start + 1)
    return max_len

arr = [1, 2, 3, 2, 4, 5, 3, 6]
print('Length of longest unique subarray:', longest_unique_subarray(arr))  # Output: 5 ([2, 4, 5, 3, 6]) 

File: javascript/arrays/longestUniqueSubarray.js
// Example: Find the length of the longest subarray with all unique elements
// Duplicates make this problem harder because you must track and skip over repeated elements efficiently

function longestUniqueSubarray(arr) {
  let maxLen = 0;
  let start = 0;
  const seen = new Map();

  for (let end = 0; end < arr.length; end++) {
    if (seen.has(arr[end]) && seen.get(arr[end]) >= start) {
      start = seen.get(arr[end]) + 1;
    }
    seen.set(arr[end], end);
    maxLen = Math.max(maxLen, end - start + 1);
  }
  return maxLen;
}

const arr = [1, 2, 3, 2, 4, 5, 3, 6];
console.log("Length of longest unique subarray:", longestUniqueSubarray(arr)); // Output: 5 ([2, 4, 5, 3, 6])
